### ISSUE TITLE
Accessibility: Decorative icons should not be read by screen readers

### ISSUE THREAD

**Issue by GuySartorelli** [silverstripe/silverstripe-admin#1957]
[WCAG 2.2 A F96](https://www.w3.org/WAI/WCAG21/Techniques/failures/F96) 2.5.3
In the left menu (and other places) we’re using a font to represent icons, and psuedo `::before` elements with CSS like `content: "C"; font-family: "silverstripe";`.

That way, the text "C" is styled to look like an icon - but obviously this could represent a problem for screen readers.

The `speak:none;` css property ([spec](https://drafts.csswg.org/css-speech-1/#speaking-props-speak)) has also been applied, though [that isn’t widely supported](https://caniuse.com/?search=speak).

We will need to change the way these icons are applied so that there's no chance of decorative icons being read as text by screen readers.

## Places I know for sure we're using fonts for icons

- in the left menu for each section
- in the "previous/next" buttons in an edit form in a gridfield
- in buttons (e.g. "save", "publish")
- in the site tree (to indicate page types)
- asset admin (e.g. edit folder button in breadcrumbs)

There's probably a ton more

## Some options

If the element with the `font-icon-*` CSS class _only_ contains an icon, it's a lot easier to deal with than an element that has an icon _and other text_ inside it.

1. Replace `::before` pseudo element with a `background-image` or other CSS that displays the icon and doesn't inject text into the DOM
2. Add `aria-hidden="true"` property to any element that has a `font-icon-*` CSS class

If the element with the `font-icon-*` CSS class contains other content beyond just the icon, the content should be moved to be _next to_ the icon element, so that one of the above (or other) mitigations can be applied.

## Notes

- The existing functionality needs to keep working in case other developers are relying on it for their own icon purposes.

## Acceptance criteria

- [ ] When purely presentational icons are displayed in the CMS, screen readers ignore them entirely
- [ ] When non-presentational icons are displayed in the CMS, screen readers read appropriate text to describe them
- [ ] The changelog clearly calls out any changes that might affect custom css/js

## PRs - Templates and simple CSS

[**kitchen sink CI run**](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/17366430816)

> [!NOTE]
> This is the first load of PRs - I'm breaking it up so it's not as much of a pain to review.
> This set of PRs includes any trivial `content:` in css that sets font icons, and any font icons set in templates.

I noticed some elements moved around by a few pixels. Some of those sorts of changes were easy to diagnose so I resolved them - but others were more effort than it's worth. We don't need pixel-perfect placement precision.

- https://github.com/silverstripe/silverstripe-admin/pull/1988
- https://github.com/silverstripe/silverstripe-framework/pull/11843
- https://github.com/silverstripe/silverstripe-cms/pull/3112
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1608
- https://github.com/silverstripe/silverstripe-login-forms/pull/228
- https://github.com/silverstripe/silverstripe-linkfield/pull/407
- https://github.com/silverstripe/silverstripe-gridfieldqueuedexport/pull/146
- https://github.com/silverstripe/silverstripe-sharedraftcontent/pull/294
- https://github.com/silverstripe/silverstripe-gridfieldextensions/pull/456
- https://github.com/silverstripe/silverstripe-advancedworkflow/pull/610
- https://github.com/tractorcow-farm/silverstripe-fluent/pull/1006
- https://github.com/silverstripe/silverstripe-gridfield-bulk-editing-tools/pull/348
- https://github.com/silverstripe/developer-docs/pull/811

## PRs - JS icons

> [!NOTE]
> This is the second load of PRs - I'm breaking it up so it's not as much of a pain to review.
> This set of PRs includes any `font-icon-*` css class in elements added through JavaScript including react.

[**Kitchen sink CI run**](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/18206494131)
Compare with a raw [`6` branch CI run](https://github.com/silverstripe/recipe-kitchen-sink/actions/runs/18177917673)

- https://github.com/silverstripe/silverstripe-admin/pull/2009
- https://github.com/silverstripe/silverstripe-linkfield/pull/411
- https://github.com/silverstripe/silverstripe-campaign-admin/pull/373
- https://github.com/silverstripe/silverstripe-mfa/pull/621
- https://github.com/silverstripe/silverstripe-gridfieldqueuedexport/pull/147
- https://github.com/silverstripe/silverstripe-versioned-admin/pull/427
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1613
- https://github.com/silverstripe/silverstripe-sharedraftcontent/pull/295
- https://github.com/silverstripe/silverstripe-gridfieldextensions/pull/458
- https://github.com/silverstripe/silverstripe-elemental/pull/1376
- https://github.com/silverstripe/developer-docs/pull/822

## PRs - PHP icons

> [!NOTE]
> This is the third load of PRs - I'm breaking it up so it's not as much of a pain to review.
> This set of PRs includes any `font-icon-*` css class in elements added through PHP including `FormAction` and `GridField_FormAction`

[**Kitchen sink CI run**](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/18480451603)

- https://github.com/silverstripe/silverstripe-framework/pull/11862
- https://github.com/silverstripe/silverstripe-cms/pull/3119
- https://github.com/silverstripe/silverstripe-admin/pull/2013
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1617
- https://github.com/silverstripe/silverstripe-versioned/pull/527
- https://github.com/silverstripe/silverstripe-versioned-admin/pull/428
- https://github.com/silverstripe/silverstripe-userforms/pull/1410
- https://github.com/silverstripe/silverstripe-linkfield/pull/416
- https://github.com/silverstripe/silverstripe-lumberjack/pull/205
- https://github.com/silverstripe/silverstripe-gridfieldqueuedexport/pull/148
- https://github.com/silverstripe/silverstripe-elemental/pull/1381
- https://github.com/silverstripe/silverstripe-gridfield-bulk-editing-tools/pull/349
- https://github.com/silverstripe/silverstripe-advancedworkflow/pull/611
- https://github.com/silverstripe/silverstripe-gridfieldextensions/pull/459
- https://github.com/silverstripe/silverstripe-queuedjobs/pull/486
- https://github.com/tractorcow-farm/silverstripe-fluent/pull/1009
- https://github.com/silverstripe/silverstripe-htmleditor-tinymce/pull/27
- https://github.com/silverstripe/developer-docs/pull/826

## PRs - modals

There's a couple of things going on in this set:

1. Hide font-icon in the modal close button
    - react modal
    - grid field import button modal
    - the old jquery UI modal
2. Make the old jquery UI modal header/close button look closer to the bootstrap one (an improvement beyond https://github.com/silverstripe/silverstripe-admin/issues/1914)

[**Kitchen sink CI run**](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/18732088572)

- https://github.com/silverstripe/webpack-config/pull/87
- https://github.com/silverstripe/silverstripe-admin/pull/2026
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1624
- https://github.com/silverstripe/silverstripe-framework/pull/11886
- https://github.com/silverstripe/silverstripe-advancedworkflow/pull/612
- https://github.com/silverstripe/silverstripe-gridfieldextensions/pull/461

## PRs - all the rest

> [!NOTE]
> This is the last load of PRs.
>
> If there are any font-icons remaining that aren't on an element with either `aria-hidden="true"` or a `aria-label`, and which aren't inside a chosen-js dropdown, this is the time to find them and report them so I can deal with them.

[**Kitchen sink CI run**](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/18767298427)
PHPUnit failure is [also happening on the `6` branch](https://github.com/silverstripe/recipe-kitchen-sink/actions/runs/18765431015)

- https://github.com/silverstripe/silverstripe-admin/pull/2024
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1623
- https://github.com/silverstripe/silverstripe-framework/pull/11885
- https://github.com/silverstripe/silverstripe-cms/pull/3125
- https://github.com/silverstripe/silverstripe-versioned/pull/528
- https://github.com/silverstripe/silverstripe-campaign-admin/pull/375
- https://github.com/silverstripe/silverstripe-versioned-admin/pull/429
- https://github.com/silverstripe/silverstripe-elemental/pull/1382

**Comment by GuySartorelli** [comment]
Some of these will be harder than others and should be done in a separate PR run if not a separate card.


## Non-trivial to update

- Any `GridField_FormAction` that has the icon added through `addExtraClass()`
  - Of particular note is `.grid-field__sort` which sets 3 different icons through css `content:` directly depending on how the item is sorted
- Any `FormAction` that has the icon added through `addExtraClass()`
  - Probably for this and the gridfield variant we can add a new `setIcon()` method, and then conditionally add an icon span in the template.
- Icons in `CMSMain_SubTree.ss` and `CMSMain_TreeNode.ss` (will be done in https://github.com/silverstripe/silverstripe-cms/issues/3101)

## Markup is managed by some third-party JS library

Maybe we edit the JS library to add the span if it's already a modified version?
Alternatively we could use `@include icon-sprites; @include sprite` or similar to add an image instead of using a font icon.

- Bootstrap modal close button
  - Uses `.modal__close-button:before, .modal-header .btn-close:before` since we can't add our own HTML markup for the js modals.
- `.popover-actions-simulate .ui-tabs-anchor` - the `ui-tabs-anchor` css class is added through jQuery UI.
- `.notice-item-close` is managed by `jquery.notice.js`

### weird interplay between templates and chosen.js means we can't just add the span

- `LeftAndMain_SilverStripeNavigator.ss`
- `LeftAndMain_ViewModeSelector.ss` (in both admin and sharedraftcontent) - note when this is resolved, need to update the selector for `$('.cms-preview .result-selected').hasClass('font-icon-columns')` in `LeftAndMain.Preview.js`
- `.bulk-actions` from `BulkActions.scss` in asset-admin
- `.gallery__sort` from `Gallery.scss` in asset-admin

**Comment by emteknetnz** [comment]
All these PRs look fine, though I had couple of bits of minor feedback on the docs PR

I'm not quite sure how much is supposed to change in this lot of PRs vs follow up PRs - on my local with all these PRs - a very quick check:
- In the left hand nav the AssetAdmin ModelAdmin icon still has a `::before` pseudo element with `content: "4"`
- In AssetAdmin, the file upload icon `::before` pseudo element has `content: "b"`

Though you've listed PHP and JS files as "TODO" (though some PRs do have JS files updated), though I'm going to assume that for now we don't need to ensure that all everything has been resolved at this stage, though they will be in follow up PRs

**Comment by GuySartorelli** [comment]
>  In the left hand nav the AssetAdmin ModelAdmin icon still has a ::before pseudo element with content: "4"

Yes but now it's inside a child `<span>` with `aria-hidden="true"`, as with all the others.

> In AssetAdmin, the file upload icon ::before pseudo element has content: "b"

The `font-icon-upload` CSS class for that element is being added through JS rather than in a template, so that'll be part of the next round of PRs.

> Though you've listed PHP and JS files as "TODO" (though some PRs do have JS files updated), though I'm going to assume that for now we don't need to ensure that all everything has been resolved at this stage, though they will be in follow up PRs

Correct - the only things that I have resolved so far are:
- `font-icon-*` CSS classes added in templates
- `content:` properties in CSS (which sometimes had to be resolved by updating markup in JS files)

**Comment by GuySartorelli** [comment]
Font icons used in the options inside a `chosen-js` dropdown are too difficult to change. I managed to get them added to child elements, but then after opening the dropdown and pressing any key other than up or down would remove the icons again.

That's because the search input field (which exists even if you don't enable searching in the dropdown) is what controls _all_ keyboard events in the dropdown. Pressing a key other than up or down triggers the search functionality, which rebuilds the options in the DOM which doesn't include the icons. There's no event fired for this so I can't hook into the event and add the icons back. I also couldn't get an event handler in earlier than the chosen-js handler to cancel those events because `vendor.js` loads before `bundle.js`.

Because of this, I won't be doing anything to the font icons which live inside options in chosen-js dropdowns.
Instead I will open a card for a future major release to stop using chosen-js (turns out if has a plethora of accessibility problems)

**Comment by GuySartorelli** [comment]
> `.notice-item-close` is managed by `jquery.notice.js`

Turns out that's not actually used, at least not anywhere I could find. The regular toasts seem to be using some other mechamism for the close buttons. So we don't need to worry about that. I'll leave the CSS there since someone may be using it somehow - don't wanna break things unnecessarily.
